caddyhttp/push/link_parser.go

```diff
@@ -0,0 +1,70 @@
+package push
+
+import (
+	"regexp"
+	"strings"
+)
+
+const (
+	commaSeparator     = ","
+	semicolonSeparator = ";"
+	equalSeparator     = "="
+)
+
+var (
+	resourceRegexp = regexp.MustCompile("^<(.*?)>;?(.*)$")
```

Comment:
For what it's worth, I wrote a simple function (not thoroughly tested, but I did get it working with your test case above) that extracts the text between `< >` (but not the other properties; those can also be parsed easily enough). Here's how it performs compared to FindAllStringSubmatch:

```
BenchmarkLinkRegex-8     	  200000	      6334 ns/op
BenchmarkLinkStrings-8   	 5000000	       332 ns/op
```

Performance is not by any means a halter for merging something like this -- but the work is half done now. :smile: Do you think we should finish it up? Here is the test/bench file:

```go
package main

import (
	"regexp"
	"strings"
	"testing"
)

var resourceRegexp = regexp.MustCompile("^<(.*?)>;?(.*)$")

var link = `</assets/css/screen.css?v=5fc240c512>; rel=preload; as=style,</content/images/2016/06/Timeouts-001.png>; rel=preload; as=image,</content/images/2016/06/Timeouts-002.png>; rel=preload; as=image")`

func BenchmarkLinkRegex(b *testing.B) {
	for n := 0; n < b.N; n++ {
		resourceRegexp.FindAllStringSubmatch(link, -1)
	}
}

func TestLinkString(t *testing.T) {
	var items []string
	end := 0
	for {
		openCaret := strings.Index(link[end:], "<") + end + 1
		if openCaret < end+1 {
			break
		}
		endCaret := strings.Index(link[openCaret:], ">") + openCaret
		items = append(items, link[openCaret:endCaret])
		end = endCaret + 1
	}
}

func BenchmarkLinkStrings(b *testing.B) {
	for n := 0; n < b.N; n++ {
		linkString(link)
	}
}

// this is the function that parses the Link header (it deserves a better name, and
// should probably parse the other components than just the resource path)
func linkString(link string) []string {
	var items []string
	end := 0
	for {
		openCaret := strings.Index(link[end:], "<") + end + 1
		if openCaret < end+1 {
			break
		}
		endCaret := strings.Index(link[openCaret:], ">") + openCaret
		items = append(items, link[openCaret:endCaret])
		end = endCaret + 1
	}
	return items
}
```